src/gennylib/include/gennylib/v1/GlobalRateLimiter.hpp

```diff
@@ -0,0 +1,147 @@
+        return std::chrono::duration_cast<std::chrono::nanoseconds>(nowTime).count();
+    }
+
+public:
+    explicit BaseGlobalRateLimiter(const RateSpec& rs)
+        : _burstSize(rs.operations), _rateNS(rs.per.count()){};
+
+    // No copies or moves.
+    BaseGlobalRateLimiter(const BaseGlobalRateLimiter& other) = delete;
+    BaseGlobalRateLimiter& operator=(const BaseGlobalRateLimiter& other) = delete;
+
+    BaseGlobalRateLimiter(BaseGlobalRateLimiter&& other) = delete;
+    BaseGlobalRateLimiter& operator=(BaseGlobalRateLimiter&& other) = delete;
+
+    ~BaseGlobalRateLimiter() = default;
+
+    /**
+     * Request to consume _burstSize number of tokens from the bucket. Does not block
+     * if the bucket is empty; does block while waiting for concurrent accesses to the
+     * bucket to finish.
+     *
+     * @return bool whether consume() succeeded. The caller is responsible for using an
+     * appropriate back-off strategy if this function returns false.
+     */
+    bool consume() {
+        if (!_emptiedTimeNSHasBeenInitialized) {
+            _emptiedTimeNSHasBeenInitialized = true;
+            // Allow one burst to go through when consume() is called the first time.
+            // It's fine that this section is not atomic, a best effort is sufficient.
+            _lastEmptiedTimeNS = now() - _rateNS;
+        }
+
+        int64_t startTime = now();
+
+        // The time the bucket was emptied before this consume() call.
+        int64_t curEmptiedTime = _lastEmptiedTimeNS.load();
+
+        // The time the bucket was emptied after this consume() call.
+        int64_t newEmptiedTime;
+
+        do {
+            newEmptiedTime = curEmptiedTime + _rateNS;
+
+            // If the new emptied time is in the future, the bucket is empty. Return early.
+            if (startTime < newEmptiedTime) {
+                return false;
+            }
+            // Use the "weak" version for performance at the expense of false negatives (i.e.
+            // `compare_exchange` not comparing equal when it should).
+        } while (!_lastEmptiedTimeNS.compare_exchange_weak(curEmptiedTime, newEmptiedTime));
+
+        return true;
+    }
+
+    constexpr int64_t getBurstSize() const {
+        return _burstSize;
+    }
+
+    constexpr int64_t getRate() const {
+        return _rateNS;
+    }
+
+private:
+    // Manually align _lastEmptiedTimeNS here to vastly improve performance.
+    // Lazily initialized by the first call to consume().
+    alignas(BaseGlobalRateLimiter::CacheLineSize) std::atomic_int64_t _lastEmptiedTimeNS = 0;
+
+    // Flag to determine when consume() is first called.
+    std::atomic_bool _emptiedTimeNSHasBeenInitialized = false;
+
+    // Note that the rate limiter as-is doesn't use the burst size, but it is cleaner to
+    // store the burst size and the rate together, since they're specified together in
+    // the YAML as RateSpec.
+    const int64_t _burstSize;
```

Comment:
¯\\\_(ツ)\_\/¯ no strong opinion but I'd just as well leave off the burstSize; at least don't expose the `getBurstSize()` getter since I wouldn't want anybody to build on it.

Reply:
Right now the phase loop uses `getBurstSize()` to figure out which iterations need rate limiting applied. Would you prefer storing the burstSize elsewhere?